src/lightning/fabric/strategies/fsdp.py

```diff
@@ -300,6 +303,143 @@ def clip_gradients_value(  # type: ignore[override]
             "Consider clipping by norm instead or choose another strategy!"
         )
 
+    def save_checkpoint(
+        self, path: _PATH, state: Dict[str, Union[Module, Optimizer, Any]], storage_options: Optional[Any] = None
+    ) -> None:
+        """Save model, optimizer, and other state in a checkpoint directory.
+
+        The directory will contain one file per process, with model- and optimizer shards stored per file. Additionally,
+        it creates a metadata file `meta.pt` with the rest of the user's state (only saved from rank 0).
+        """
+        if not _TORCH_GREATER_EQUAL_2_0:
+            raise NotImplementedError(
+                "Saving and loading checkpoints with the `FSDPStrategy` is not supported in PyTorch < 2.0."
+                " Please upgrade `torch` or file an issue: `https://github.com/Lightning-AI/lightning/issues`."
+            )
+        if storage_options is not None:
+            raise TypeError(
+                "`FSDPStrategy.save_checkpoint(..., storage_options=...)` is not supported because"
+                " `FSDPStrategy` does not use the `CheckpointIO`."
+            )
+
+        from torch.distributed._shard.checkpoint import FileSystemWriter, save_state_dict
+        from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+        modules = [module for module in state.values() if isinstance(module, FSDP)]
+        if len(modules) == 0:
+            raise ValueError(
+                "Could not find a FSDP model in the provided checkpoint state. Please provide the model as"
+                " part of the state like so: `save_checkpoint(..., state={'model': model, ...})`. Make sure"
+                " you set up the model (and optimizers if any) through the strategy before saving the checkpoint."
+            )
+        elif len(modules) > 1:
+            raise ValueError(
+                "Found multiple FSDP modules in the given state. Saving checkpoints with FSDP is"
+                " currently limited to a single model per checkpoint. To save multiple models, call the"
+                " save method for each model separately with a different path."
+            )
+
+        module = modules[0]
+
+        # broadcast the path from rank 0 to ensure all the states are saved in a common path
```

Comment:
Is this an absolute path? We need to ensure that all nodes can reach the path and have the same path mounted, which may not always be the case.
Maybe we can make path broadcasting opt-in? (not for this PR necessarily)

Reply:
It can be relative or absolute, it does not matter. 

> We need to ensure that all nodes can reach the path and have the same path mounted, which may not always be the case.

I don't see how that is related to broadcasting the path. If the user doesn't have a shared filesystem, then they will end up with a folder with the local states and they can merge the folders later if they want. The broadcasting is here simply to avoid issues with non-deterministic filenames, such as when the user adds a time stamp and due to multiprocessing ends up with different timestamps per rank. We could of course leave the broadcasting up to the user, but then the experience would be inconsistent as you switch from one strategy (ddp which only saves on rank 0) to FSDP/Deepspeed. 

The Lightning Trainer also broadcasts the path to avoid these race conditions.

Reply:
Sounds good @awaelchli , thanks for clarifying